### PR TITLE
Fixes issue with package-deps-hash where it does not return the entire list of file hashes

### PR DESCRIPTION
**Pull request by nickpape-msft**

### CODE REVIEW

**Comment by nickpape-msft** [comment]
@dzearing I am making some changes to rush that exposed this bug.

File: package-deps-hash/src/getPackageDeps.ts [hunk 1]
```diff
@@ -11,63 +11,61 @@ export function getPackageDeps(packagePath: string = process.cwd(), excludedPath
...
+      files: {}
+    };
+    const processOutputBlocks: string[] = stdout.split(PROCESS_OUTPUT_DELIMITER);
 
-        if (processOutputBlocks[1]) {
-          const filesToHash: string[] = [];
+    // Note: The output of git ls-tree uses \n newlines regardless of OS.
+    processOutputBlocks[0].split('\n').forEach(line => {
+      if (line) {
+        const parts: string[] = line.substr(line.indexOf('blob ') + 5).split('\t');
```
Comment (iclanton) [inline package-deps-hash/src/getPackageDeps.ts]: Can you give an example of what a line might look like? This string manipulation is a little cryptic without an example.

File: package-deps-hash/src/getPackageDeps.ts [hunk 2]
```diff
@@ -11,63 +11,61 @@ export function getPackageDeps(packagePath: string = process.cwd(), excludedPath
...
-            .trim()
-            .split('\n')
-            .forEach(line => {
-              const parts: string[] = line.trim().split(' ');
+      // Note: The output of git hash-object uses \n newlines regardless of OS.
+      processOutputBlocks[1]
+        .trim()
+        .split('\n')
+        .forEach(line => {
+          const parts: string[] = line.trim().split(' ');
```
Comment (iclanton) [inline package-deps-hash/src/getPackageDeps.ts]: Same down here.

File: package-deps-hash/src/getPackageDeps.ts [hunk 3]
```diff
@@ -11,63 +11,61 @@ export function getPackageDeps(packagePath: string = process.cwd(), excludedPath
...
-            const parts: string[] = line.substr(line.indexOf('blob ') + 5).split('\t');
-            if (!excludedHashes[parts[1]]) {
-              changes.files[parts[1]] = parts[0];
-            }
-          }
-        });
+    const changes: IPackageDeps = {
+      files: {}
+    };
+    const processOutputBlocks: string[] = stdout.split(PROCESS_OUTPUT_DELIMITER);
```
Comment (pgonzal) [inline package-deps-hash/src/getPackageDeps.ts]: What if the processOutputBlocks[] array has only one element?  We are basically assuming that "git ls-tree" will produce a predictable output, which could be a brittle assumption.

File: package-deps-hash/src/getPackageDeps.ts [hunk 4]
```diff
@@ -11,63 +11,61 @@ export function getPackageDeps(packagePath: string = process.cwd(), excludedPath
...
 
-        if (processOutputBlocks[1]) {
-          const filesToHash: string[] = [];
+    // Note: The output of git ls-tree uses \n newlines regardless of OS.
+    processOutputBlocks[0].split('\n').forEach(line => {
+      if (line) {
+        const parts: string[] = line.substr(line.indexOf('blob ') + 5).split('\t');
+        if (!excludedHashes[parts[1]]) {
```
Comment (pgonzal) [inline package-deps-hash/src/getPackageDeps.ts]: This logic would be way more readable if you included a comment showing an example output line, and examples of what the parts[] items look like.

File: package-deps-hash/src/getPackageDeps.ts [hunk 5]
```diff
@@ -11,63 +11,61 @@ export function getPackageDeps(packagePath: string = process.cwd(), excludedPath
...
-              if (parts.length === 2) {
-                if (parts[0] === 'D') {
-                  delete changes.files[parts[1]];
-                } else {
-                  if (!excludedHashes[parts[1]]) {
-                    filesToHash.push(parts[1]);
-                  }
-                }
+          if (parts.length === 2) {
+            if (parts[0] === 'D') {
```
Comment (pgonzal) [inline package-deps-hash/src/getPackageDeps.ts]: Yeah, no idea what "D" means or what the input looks like...

File: package-deps-hash/src/getPackageDeps.ts [hunk 6]
```diff
@@ -11,63 +11,61 @@ export function getPackageDeps(packagePath: string = process.cwd(), excludedPath
   }
 
   return new Promise((complete) => {
```
Comment (pgonzal) [inline package-deps-hash/src/getPackageDeps.ts]: If everything is synchronous, why do we need this Promise boilerplate?

Comment (nickpape-msft) [inline package-deps-hash/src/getPackageDeps.ts]: We don't technically anymore. Although that will be a major version increment.

**Comment by nickpape-msft** [comment]
Pete and I discussed this last night. Most PR comments here are re source that I did not change. However, Pete convinced me that when you work on a file, that is the right time to perform code cleanup. Hence I will address some of the concerns on this PR before merging.

File: package-deps-hash/src/getPackageDeps.ts (hunk 3 above)
Comment (nickpape-msft) [inline package-deps-hash/src/getPackageDeps.ts]: So it looks like processOutputBlocks just splits the output (with the delimiter) into 2 arrays. So it should only have 1 element if the delimiter was not written, which would occur if the `git ls-tree` command returns a non-zero exit code. However, the way split works, is that the entire string goes into `processOutputBlocks[0]` if the delimiter is not found.

Either way, we check the value of `processOutputBlocks[1]` before doing anything.

I can add some comments about that.

File: package-deps-hash/src/getPackageDeps.ts (hunk 4 above)
Comment (nickpape-msft) [inline package-deps-hash/src/getPackageDeps.ts]: Done

File: package-deps-hash/src/getPackageDeps.ts (hunk 5 above)
Comment (nickpape-msft) [inline package-deps-hash/src/getPackageDeps.ts]: Added a comment

File: package-deps-hash/src/getPackageDeps.ts (hunk 2 above)
Comment (nickpape-msft) [inline package-deps-hash/src/getPackageDeps.ts]: Added some comments

File: package-deps-hash/src/getPackageDeps.ts (hunk 1 above)
Comment (nickpape-msft) [inline package-deps-hash/src/getPackageDeps.ts]: Gave an example

**Pull request merged**